designs/node-overlays.md

```diff
@@ -0,0 +1,186 @@
+
+```yaml
+kind: NodeOverlay
+metadata:
+  name: extended-resource
+spec:
+  selector:
+    matchLabels:
+      karpenter.sh/capacity-type: on-demand
+    matchExpressions:
+    - key: node.kubernetes.io/instance-type
+      operator: In
+      values: ["m5.large", "m5.2xlarge", "m5.4xlarge", "m5.8xlarge", "m5.12xlarge"]
+  capacity:
+    smarter-devices/fuse: 1
+```
+
+Add memory overhead of of 10Mi and 50Mi for all instances with more than 2Gi memory
+* https://github.com/aws/karpenter-provider-aws/issues/5161
+
+```yaml
+kind: NodeOverlay
+metadata:
+  name: memory-default
+spec:
+  overhead:
+    memory: 10Mi
+---
+kind: NodeOverlay
+metadata:
+  name: memory
+spec:
+  weight: 1
+  selector:
+    matchExpressions:
+      key: karpenter.k8s.aws/instance-memory
+      operator: Gt
+      values: [2048]
+  overhead:
+    memory: 50Mi
+---
+```
+
+### Selectors
+
+NodeOverlay's selector semantic allows for fine grained control over when values are injected into the scheduling simulation. An empty selector applies to all simulations, or 1:n with InstanceType. Selectors could be written that define a 1:1 relationship between NodeOverlay and InstanceType. Selectors can even be written to apply n:1 with InstanceType, through additional labels that constrain by NodePool, NodeClass, zone, or other scheduling dimensions.
+
+A [previous proposal](https://github.com/aws/karpenter-provider-aws/pull/2390/files#diff-b7121b2d822e57b70203794f3b6367e7173b16d84070c3712cd2f15b8dbd11b2R133) explored the tradeoffs between a 1:1 semantic vs more flexible alternatives. It raised valid observability challenges around communicating to customers exactly how the CR was modifying the scheduling simulation. It recommended a 1:1 approach, with the assumption that some system would likely dynamically codegenerate the CRs for the relevant use case. We received [direct feedback](https://github.com/aws/karpenter-provider-aws/pull/2404#issuecomment-1250688882) from some customers that the number of variations would be combinatorial, and thus unmanagable for their use case. Ultimately, we paused this analysis in favor of other priorities.
+
+After two years of additional customer feedback in this problem space, this proposal asserts that the flexibility of a selector-based approach is a signficant simplification for customers compared to a 1:1 approach and avoids the scalability challenges associated with persisting large numbers of CRs in the Kubernetes API. This decision will require us to work through the associated observability challenges.
+
+### Merging
+
+Our decision to pursue Selectors that allow for an n:m semantic presents a challenge where multiple NodeOverlays could match a simulation, but with different values. To address this, we will support an optional `.spec.weight` parameter (discussed in the [previous proposal](https://github.com/aws/karpenter-provider-aws/pull/2390/files#diff-b7121b2d822e57b70203794f3b6367e7173b16d84070c3712cd2f15b8dbd11b2R153)).
```

Comment:
I'm thinking about these overlays as a combination of mathematical operations on a subset of instance types for a subset of attributes. Is it possible to just arrange these mathematical operations in a way like PEMDAS such that we don't have to think about weight and can just merge them?  I think this would be option 2.

Reply:
This is a cool idea, but I'm not sure about the use cases it unlocks.